operator/pkg/clustermesh/client.go

```diff
@@ -0,0 +1,170 @@
+// list/watch calls to also make sure that we fake the service name refererenced
+// by EndpoinSlices and will revert those changes at the create/update level to make
+// sure that the correct data is written to the Kubernetes API.
+type meshClient struct {
+	k8sClient.Clientset
+}
+
+func (c meshClient) DiscoveryV1() discoveryv1.DiscoveryV1Interface {
+	return meshClientDiscoveryV1{c.Clientset.DiscoveryV1()}
+}
+
+type meshClientDiscoveryV1 struct {
+	discoveryv1.DiscoveryV1Interface
+}
+
+func (c meshClientDiscoveryV1) EndpointSlices(namespace string) discoveryv1.EndpointSliceInterface {
+	return meshClientEndpointSlice{c.DiscoveryV1Interface.EndpointSlices(namespace)}
+}
+
+type meshClientEndpointSlice struct {
+	discoveryv1.EndpointSliceInterface
+}
+
+func isServiceOwnerReferrence(ownerReference metav1.OwnerReference) bool {
+	return ownerReference.APIVersion == "v1" && ownerReference.Kind == "Service"
+}
+
+// addEndpointSliceMeshHacks fakes that the parent service of the EndpointSlice
+// has the cluster name in its name by changing the owner reference and the label
+// pointing to the service name. It also make sure that there is a label `mesh.cilium.io/real-service-name`
+// to store the real service name before its modification.
+func addEndpointSliceMeshHacks(endpointSlice *v1.EndpointSlice) {
+	if endpointSlice == nil || endpointSlice.Labels[discovery.LabelManagedBy] != ciliumControllerName {
+		return
+	}
+
+	endpointSlice.Labels[meshRealServiceNameLabel] = endpointSlice.Labels[discovery.LabelServiceName]
+	endpointSlice.Labels[discovery.LabelServiceName] = endpointSlice.Labels[meshRealServiceNameLabel] + "-" + endpointSlice.Labels[mcsapiv1alpha1.LabelSourceCluster]
+
+	for i, ownerReference := range endpointSlice.OwnerReferences {
+		if !isServiceOwnerReferrence(ownerReference) || ownerReference.Name != endpointSlice.Labels[meshRealServiceNameLabel] {
+			continue
+		}
+
+		endpointSlice.OwnerReferences[i].Name = endpointSlice.Labels[discovery.LabelServiceName]
+	}
+}
+
+// removeEndpointSliceMeshHacks revert the change to the parent service made in
+// addEndpointSliceMeshHacks.
+func removeEndpointSliceMeshHacks(endpointSlice *v1.EndpointSlice) {
+	if endpointSlice == nil || endpointSlice.Labels[discovery.LabelManagedBy] != ciliumControllerName {
+		return
+	}
+
+	for i, ownerReference := range endpointSlice.OwnerReferences {
+		if !isServiceOwnerReferrence(ownerReference) || ownerReference.Name != endpointSlice.Labels[discovery.LabelServiceName] {
+			continue
+		}
+
+		endpointSlice.OwnerReferences[i].Name = endpointSlice.Labels[meshRealServiceNameLabel]
+	}
+
+	endpointSlice.Labels[discovery.LabelServiceName] = endpointSlice.Labels[meshRealServiceNameLabel]
+	delete(endpointSlice.Labels, meshRealServiceNameLabel)
+}
+
+func (c meshClientEndpointSlice) Create(ctx context.Context, endpointSlice *v1.EndpointSlice, opts metav1.CreateOptions) (*v1.EndpointSlice, error) {
+	removeEndpointSliceMeshHacks(endpointSlice)
```

Comment:
AFAIK the return value of the Create/Update are used instead of what's given to the Update/Create so this should be fine, but I need to double check... Could be related to the issues you pointed out when testing it...

Reply:
Yes so the calls are here:
- https://github.com/cilium/endpointslice/blob/master/reconciler.go#L383
- https://github.com/cilium/endpointslice/blob/master/reconciler.go#L368

And those calls are done pretty much at the very end of the reconciler + use the return value rather than the arg (except if the error is not nill but it looks fine).